// ndp/deployNdp.js
const Deploy = require('./interface/Deploy');
const ServerInfo = require('./interface/ServerInfo');
const CancelBuild = require('./interface/CancelBuild');
const Cluster = require('./interface/Cluster');
const Env = require('./tools/Env');
const failRetryWrapper = require('./tools/failRetryWrapper');
const checkPackageVersion = require('./tools/checkPackageVersion');
const checkNodeVersion = require('./tools/checkNodeVersion');
const getCurBranch = require('./tools/get_cur_branch');

const deployNdp = async apiEnv => {
  checkNodeVersion();
  await checkPackageVersion();
  const env = new Env(apiEnv);
  const deploy = new Deploy({
    cluster: new Cluster({ env }),
    serverInfo: new ServerInfo({ env }),
    cancelBuild: new CancelBuild({ env }),
    env,
  });
  const retryDeploy = failRetryWrapper({
    fn: deploy.deploy.bind(deploy),
    maxRetry: Number(env.getEnv('NDP_RETRY')),
    retryInterval: Number(env.getEnv('NDP_INTERVAL')),
  });
  let branch = env.getEnv('NDP_BRANCH');
  if (branch === 'HEAD') {
    branch = await getCurBranch();
  }
  await retryDeploy({
    clusterId: env.getEnv('NDP_CID'),
    branch,
    hash: env.getEnv('NDP_HASH'),
  });
};
module.exports = deployNdp;
