// ndp/rollbackNdp.js
const RollBack = require('./interface/RollBack');
const BuildHistory = require('./interface/BuildHistory');
const ServerInfo = require('./interface/ServerInfo');
const Cluster = require('./interface/Cluster');
const Env = require('./tools/Env');
const ask = require('./tools/ask');
const checkPackageVersion = require('./tools/checkPackageVersion');
const checkNodeVersion = require('./tools/checkNodeVersion');

const rollBackNdp = async apiEnv => {
  checkNodeVersion();
  await checkPackageVersion();
  const env = new Env(apiEnv);
  const clusterId = env.getEnv('NDP_CID');
  const buildHistory = new BuildHistory({
    env,
  });
  const rollBack = new RollBack({
    env,
    serverInfo: new ServerInfo({ env }),
    cluster: new Cluster({ env }),
  });
  const list = await buildHistory.list(clusterId);
  const versionId = await ask({
    type: 'rawlist',
    name: 'versionid',
    choices: list.map(({ buildTime, branch, hash, id }) => ({
      value: id,
      name: `构建时间:${buildTime} 分支:${branch} 哈希:${hash}`,
    })),
    default: 0,
    message: '选择回滚版本',
  });
  await rollBack.back({ clusterId, versionId });
};
module.exports = rollBackNdp;
